### ISSUE TITLE
noqa E501 not being respected for some cases of long lines

### ISSUE THREAD

**Issue by LefterisJP** [charliermarsh/ruff#613]
I have noticed that `noqa: E501` is not respected for some long line cases. I can't really understand which cases that is. For `line-length=99` though this one for example is an interesting one:

```python
test1 = "this is a long line that should have hit us with E501 but thanks to the noqa it does not hit us at all in stark contract with the stuff below"  # noqa: E501

test2 = {
        'from_address': 'foo',
        'pubkey': 'this seems to actually hit the problem and not respect the E501 error noqa keyword',  # noqa: E501
        'withdrawal_credentials': 'Here the noqa is respected for some reason and ruff will not complain at all',  # noqa: E501
}
```

This raises an error in line 5
```
Found 1 error(s).
test.py:5:1: E501 Line too long (117 > 99 characters)
```

While E501 should have been respected there too.

What's even more weird is that if I remove `'from_address': 'foo',` from there it all works fine

The following raises no errors

```python
test1 = "this is a long line that should have hit us with E501 but thanks to the noqa it does not hit us at all in stark contract with the stuff below"  # noqa: E501

test2 = {
        'pubkey': 'this seems to actually hit the problem and not respect the E501 error noqa keyword',  # noqa: E501
        'withdrawal_credentials': 'Here the noqa is respected for some reason and ruff will not complain at all',  # noqa: E501
}
```

**Comment by charliermarsh** [comment]
Thanks! I think I have an idea of what’s going on here… There’s some special-casing around strings to allow noqa comments to come at the end of a multi-line string, and thinking back to the code, I may have assumed there’s never more than one string to per line (which is obviously wrong). Will fix soon, sorry about that.

**Comment by charliermarsh** [comment]
(I’m guessing that it’s leading to an off-by-one or off-by-N error where N is the number of lines with > 1 string.)

**Comment by LefterisJP** [comment]
If I could run a test version I could see all those cases.

Or you could also probably easily reproduce them all.
I am working on rotki: https://github.com/rotki/rotki/ and we utilize flake8 but still got plenty of noqa around the code.

With a super simple pytoml setting line length to 99 you can see all of them.

```
[tool.ruff]
line-length = 99

# Enable Flake's "E" and "F" codes by default.
select = ["E", "F"]
ignore = [
    "E402",  # module level import at file top. https://www.flake8rules.com/rules/E402.html
    "N818",  # error suffix in exception names
]	   
```

**Comment by charliermarsh** [comment]
Awesome. I went ahead and ran with #615, and I only saw a few E501 errors.

Unlike Flake8, for multiline strings, Ruff _requires_ that the E501 come at the end. So to fix those errors, e.g. in `gevent.py`, modify the docstring like:

```py
@contextmanager
def savepoint_ctx(
        self,
        savepoint_name: Optional[str] = None,
) -> Generator['DBCursor', None, None]:
    """
    Creates a savepoint context with the provided name. If the code inside the savepoint fails,
    rolls back this savepoint, otherwise releases it (aka forgets it -- this is not commited to the DB).
    Savepoints work like nested transactions, more information here: https://www.sqlite.org/lang_savepoint.html
    """  # noqa: E501
```

**Comment by charliermarsh** [comment]
Let me know if you run into any other issues! Will cut a release (v0.0.103) now with that fix.

**Comment by charliermarsh** [comment]
(Separately: I recently released [`flake8-to-ruff`](https://pypi.org/project/flake8-to-ruff/) which can help with automatic conversion of Flake8 `setup.cfg` to Ruff `pyproject.toml`. Would love any feedback!)

**Comment by LefterisJP** [comment]
> (Separately: I recently released [`flake8-to-ruff`](https://pypi.org/project/flake8-to-ruff/) which can help with automatic conversion of Flake8 `setup.cfg` to Ruff `pyproject.toml`. Would love any feedback!)

Thank you so much charlie. This fixed all of our cases. For the others ones you mentioned I applied your suggested fix.

As for flake8-to-ruff it works fine. Just tells me some warnings are not yet supported by ruff. It's the B and W ones. One is bugbear and other is general warnings.

would love for them to be also supported and then we can get rid of flake8, while for now I guess I will run them in parallel. ruff is so fast it makes no difference :smile:
